File: rsvc/src/model/agent.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Agent {
    name: String,
    security_profile: String,
    id: String,
    e_tag: String,
    entity_id: String
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateAgentRequest {
    name: String,
    security_profile: String,
    entity_id: String
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateAgentRequest {
    name: String,
    security_profile: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OnlineStatus {
    status: String,
    since: String
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDataSourceConfigurationRequest {
    configuration_id: String,
    data_sources: Vec<DataSource>
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DataSourceConfiguration {
    id: String,
    e_tag: String,
    configuration_id: String,
    data_sources: Vec<DataSource>
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DataSource {
    name: String,
    description: String,
    data_points: Vec<DataPoint>,
    custom_data: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DataPoint {
    id: String,
    name: String,
    description: String,
    #[serde(rename = "type")]
    typed: String,
    unit: String,
    #[serde(rename = "camelCase")]
    custom_data: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TokenKey {
    alg: String,
    #[serde(rename = "use")]
    used: String,
    value: String,
    e: String,
    n: String,
    kty: String,
    kid: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AccessToken {
    access_token: String,
    token_type: String,
    expires_in: String,
    jti: String,
    scope: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Configuration {
    content: OnboardingConfigurationContent,
    expiration: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OnboardingConfigurationContent {
    base_url: String,
    iat: String,
    client_credential_profile: String,
    client_id: String,
    tenant: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RotationKeys {
    client_id: String,
    jwks: Jwks
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Jwks {
    keys: Vec<Key>
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Key {
    e: String,
    n: String,
    kty: String,
    kid: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ClientIdentifier {
    client_id: String,
    client_secret: String,
    client_secret_expires_at: String,
    grant_types: Vec<String>,
    registration_access_token: String,
    registration_client_uri: String,
    token_endpoint_auth_method: String
}

File: gsvc/pkg/model/agent.go
package model

import (
	"github.com/go-openapi/strfmt"
)

type CreateAgentRequest struct {
	Name            string `json:"name"`
	SecurityProfile string `json:"securityProfile"`
	EntityId        string `json:"entityId"`
}

type Agent struct {
	Name            string `json:"name"`
	SecurityProfile string `json:"securityProfile"`
	Id              string `json:"id"`
	ETag            string `json:"eTag"`
	EntityId        string `json:"entityId"`
}

type PagedAgent struct {
	Content          []Agent `json:"content"`
	Last             bool    `json:"last"`
	TotalPages       int     `json:"totalPages"`
	TotalElements    int     `json:"totalElements"`
	NumberOfElements int     `json:"numberOfElements"`
	First            bool    `json:"first"`
	Sort             string  `json:"sort"`
	Size             int     `json:"size"`
	Number           int     `json:"number"`
}

type Order struct {
	Direction    string `json:"direction"`
	Property     string `json:"property"`
	IgnoreCase   bool   `json:"ignoreCase"`
	NullHandling string `json:"nullHandling"`
	Descending   bool   `json:"descending"`
	Ascending    bool   `json:"ascending"`
}

type UpdateAgentRequest struct {
	Name            string `json:"name"`
	SecurityProfile string `json:"securityProfile"`
}

type OnlineStatus struct {
	Status string           `json:"status"`
	Since  *strfmt.DateTime `json:"timestamp"`
}

type UpdateDataSourceConfigurationRequest struct {
	ConfigurationId string       `json:"configurationId"`
	DataSources     []DataSource `json:"dataSources"`
}

type DataSourceConfigution struct {
	Id              string       `json:"id"`
	ETag            string       `json:"eTag"`
	ConfigurationId string       `json:"configurationId"`
	DataSources     []DataSource `json:"dataSources"`
}

type DataSource struct {
	Name        string      `json:"name"`
	Description string      `json:"description"`
	DataPoints  []DataPoint `json:"dataPoints"`
	CustomData  string      `json:"customData"`
}

type DataPoint struct {
	Id          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Type        string `json:"type"`
	Unit        string `json:"unit"`
	CustomData  string `json:"customData"`
}

type Configuration struct {
	Content    OnboardingConfigurationContent `json:"id"`
	Expiration *strfmt.DateTime               `json:"timestamp"`
}

type OnboardingConfigurationContent struct {
	BaseUrl                 string `json:"baseUrl"`
	Iat                     string `json:"iat"`
	ClientCredentialProfile string `json:"clientCredentialProfile"`
	ClientId                string `json:"clientId"`
	Tenant                  string `json:"tenant"`
}

type OnboardingStatus struct {
	Status string `json:"status"`
}

type RegisterAgentRequest struct {
	Jwks Jwkeys `json:"jwks"`
}

type Jwkeys struct {
	Keys []Key `json:"keys"`
}

type Key struct {
	E   string `json:"e"`
	N   string `json:"n"`
	Kty string `json:"kty"`
	Kid string `json:"kid"`
}

type RotationKeys struct {
	ClientId string `json:"client_id"`
	Jwks     Jwkeys `json:"jwks"`
}

type ClientIdentifier struct {
	ClientId                string `json:"client_id"`
	ClientSecret            string `json:"client_secret"`
	ClientSecretExpiresAt   string `json:"client_secret_expires_at"`
	GrantTypes              string `json:"grant_types"`
	RegistrationAccessToken string `json:"registration_access_token"`
	RegistrationClientUri   string `json:"registration_client_uri"`
	TokenEndpointAuthMethod string `json:"token_endpoint_auth_method"`
}

type AccessTokenRequest struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	ExpiresIn   string `json:"expires_in"`
	Jti         string `json:"jti"`
	Scope       string `json:"scope"`
}

type TokenKey struct {
	Alg   string `json:"alg"`
	Use   string `json:"use"`
	Value string `json:"value"`
	E     string `json:"e"`
	N     string `json:"n"`
	Kty   string `json:"kty"`
	Kid   string `json:"kid"`
}

type TokenKeys struct {
	Keys []TokenKey `json:"keys"`
}

type UnAuthorized struct {
	Id      string `json:"id"`
	Message string `json:"message"`
}
